### PR TITLE
Adds a demo app to help test and facilitate Firebase Auth API development

### PR DESCRIPTION
**Pull request by bojeil-google**
Adds a new directory in auth package demo.
3 main steps are needed:

1. config.js needs to be set up in `demo/public`.
2. Run in `demo/`: `firebase use --add`
3. `cd ..` and `yarn run demo`

### CODE REVIEW

File: packages/auth/buildtools/run_demo.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#
+# Usage:
+# $ buildtools/run_demo.sh
+
+# CD to the root packages/auth directory, which should be the parent directory of
+# buildtools/.
+cd "$(dirname $(dirname "$0"))"
+# Go back to repo root and build all binaries needed for the demo app.
+cd ../..
+yarn
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply packages/auth/buildtools/run_demo.sh]: Done.

File: packages/auth/buildtools/run_demo.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+cd packages/auth
+# Make dist directory if not already there.
+mkdir -p demo/public/dist
+# Copy app, auth and database binaries to demo dist directory.
+cp ../firebase/firebase-app.js demo/public/dist/firebase-app.js
+cp ../firebase/firebase-auth.js demo/public/dist/firebase-auth.js
+cp ../firebase/firebase-database.js demo/public/dist/firebase-database.js
+# Serve demo app.
+cd demo
+firebase serve
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply packages/auth/buildtools/run_demo.sh]: Done.

**Comment by jshcrowthe** [comment]
> **Can we leverage the app configured at the top level for testing?**
I am open to that. I wanted to contain and isolate Auth changes so I don't affect other Firebase services. We can follow up with enhancements on this after running this by the other owners.

Your proposed config will not break anything in the other tests (all of the other tests don't touch the `users` property you configure), so I would say lets leverage existing config here first as it complicates the SDK development workflow to have two separate project config steps. But if you can get this in, in the next couple weeks I'm fine going ahead.

> **How do we do incremental builds?**
I am not sure what you mean by incremental builds. Do you mean building and deploying locally multiple versions of the test app?

In your demo script, you rebuild the _entire_ SDK before you run the demo. It's just quite costly and you could implement something like webpack-dev-server (see https://github.com/firebase/firebase-js-sdk/blob/master/packages/firebase/package.json#L23) to simplify this and close the feedback loop.

**Comment by bojeil-google** [comment]
Ok, incremental builds sound like a good feature to have. Let's log an issue to track that and implement it separately. Also I think we can move the hosting config to the root package and leverage `yarn test:setup`. We'll need to remove the rule: ` ".read": "auth != null", ".write": "auth != null"`. I don't think I use it so we should be fine without it.

**Pull request merged**